### DISCUSSION TITLE
Integration tests and where their location

### DISCUSSION THREAD

**Discussion by tsvobo** [benbjohnson/wtf#39] [Q&A] [answered]
Hi Ben,
I'd like to ask about the possibilities of adding integration tests following WTF layout.

Based on the WTF code, it looks like there are unit tests and end-to-end tests of the commands (`cmd`) which test the whole app including its bootstrap.

I'd like to add some integration tests that would cover only some of the layers (not the whole E2E feature). Let's say there is a service func with business logic (in `wtf` package) that internally orchestrates http client requests (`http` package) and calls to DB (`sqllite` package).
On the first though it would make sense to colocate it with the actual business logic that is covered by the integration test but that would end up in a cyclic dependency of those 3 packages. The second possible location could be `cmd` (or some of its sub-package), however, I have a feeling that it could be confusing as it would mix E2E tests covering the whole app and some partial integration tests. Last option that comes to my mind is having a new package, e.g. `it`, and place all the integration tests there.

What do you think about it?

Thank you for the clarification.
Tomáš

**Comment by benbjohnson** [comment, marked as answer]
I like using `_test` packages that are colocated with one of the layers I'm testing. That adds an additional package to prevent cyclic dependencies but it doesn't add a new directory.

For example, you can have a test in `http/http_integration_test.go`:

```go
package http_test

import (
	"testing"

	"github.com/benbjohnson/wtf/http"
	"github.com/benbjohnson/wtf/sqlite"
)

func TestIntegration(t *testing.T) {
	server := http.NewServer()
	db := http.NewDB()
	...
}
```

**Reply by tsvobo** [reply]
I see, that looks like a better approach, thank you!
